openwisp_monitoring/check/classes/iperf.py

```diff
@@ -0,0 +1,180 @@
+        # because it may be possible that authentication (publickey) failed.
+        # or any other failure happened during connect().
+        if not device_connection.is_working:
+            logger.warning(f'{device}: SSH connection is not working, Iperf skipped!')
+            return
+        servers = self._get_iperf_servers(device.organization.id)
+        command = f'iperf3 -c {servers[0]} -J'
+        res, exit_code = device_connection.connector_instance.exec_command(
+            command, raise_unexpected_exit=False
+        )
+        if store and exit_code != 0:
+            self.store_result_fail()
+            device_connection.disconnect()
+            return
+        else:
+            result_dict_tcp = self._get_iperf_result(res, mode='TCP')
+            # UDP
+            command = f'iperf3 -c {servers[0]} -u -J'
+            res, exit_code = device_connection.connector_instance.exec_command(
+                command, raise_unexpected_exit=False
+            )
+            if store and exit_code != 0:
+                self.store_result_fail()
+                device_connection.disconnect()
+                return
+            else:
+                result_dict_udp = self._get_iperf_result(res, mode='UDP')
+                if store:
+                    self.store_result({**result_dict_tcp, **result_dict_udp})
+            device_connection.disconnect()
+            return {**result_dict_tcp, **result_dict_udp}
+
+    def _get_device_connection(self, device):
+        """
+        Returns an active SSH DeviceConnection for a device.
+        """
+        openwrt_ssh = UPDATE_STRATEGIES[0][0]
+        device_connection = DeviceConnection.objects.filter(
+            device_id=device.id,
+            update_strategy=openwrt_ssh,
+            enabled=True,
+            is_working=True,
+        ).first()
+        return device_connection
+
+    def _get_iperf_servers(self, organization_id):
+        """
+        Get iperf test servers
+        """
+        org_servers = app_settings.IPERF_SERVERS.get(str(organization_id))
+        return org_servers
+
+    def _get_iperf_result(self, res, mode):
+        """
+        Get iperf test result
+        """
+        res_dict = json.loads(res)
+        if mode == 'TCP':
+            # Gbps = Gigabits per second
+            # GB = GigaBytes
+            sent_json = res_dict['end']['sum_sent']
+            recv_json = res_dict['end']['sum_received']
+            sent_bytes = sent_json['bytes']
+            sent_bytes_GB = sent_bytes / 1000000000
+            sent_bps = sent_json['bits_per_second']
+            sent_Gbps = sent_bps / 1000000000
+            received_bytes = recv_json['bytes']
+            received_bytes_GB = received_bytes / 1000000000
+            received_bps = recv_json['bits_per_second']
+            received_Gbps = received_bps / 1000000000
+            retransmits = sent_json['retransmits']
+
+            result = {
+                'iperf_result': 1,
```

Comment:
I'm also thinking on this, I will do some research and try to implement different logic :+1:

Reply:
Opened : https://github.com/openwisp/openwisp-monitoring/issues/405

Reply:
> I'm also thinking on this, I will do some research and try to implement different logic +1

Can we do something like if any (tcp, udp) of iperf get passed, we can store `iperf_result` field to 1 otherwise 0. For eg: 
```bash
# If both test passed
'values': {'iperf_result': 1, 'sent_bps': 1.49, 'received_bps': 1.49, 'sent_bytes': 1.87, 'received_bytes': 1.87, 'retransmits': 0, 'jitter': 0.01, 'packets': 898, 'lost_packets': 0, 'lost_percent': 0.0},
# TCP passed UDP failed (previously all values was set to '0'), Now :
'values': {'iperf_result': 1, 'sent_bps': 1.5, 'received_bps': 1.5, 'sent_bytes': 1.87, 'received_bytes': 1.87, 'retransmits': 0, 'jitter': 0.0, 'packets': 0, 'lost_packets': 0, 'lost_percent': 0.0}
# Both fails
'values': {'iperf_result': 0, 'sent_bps': 0.0, 'received_bps': 0.0, 'sent_bytes': 0.0, 'received_bytes': 0.0, 'retransmits': 0, 'jitter': 0.0, 'packets': 0, 'lost_packets': 0, 'lost_percent': 0.0}
```